clang/lib/Lex/Lexer.cpp

```diff
@@ -3183,8 +3193,35 @@ bool Lexer::LexEndOfFile(Token &Result, const char *CurPtr) {
       DiagID = diag::ext_no_newline_eof;
     }
 
-    Diag(BufferEnd, DiagID)
-      << FixItHint::CreateInsertion(EndLoc, "\n");
+    if (LastNewline.empty()) {
+      Diag(BufferEnd, DiagID) << FixItHint::CreateInsertion(EndLoc, "\n");
+    } else {
+      // While the file physically ends in a newline, the previous
+      // line might have ended in a splice, so it would be deleted
+      StringRef WithoutLastNewline =
+          StringRef(BufferStart, LastNewline.data() - BufferStart);
+      while (!WithoutLastNewline.empty()) {
+        if (isHorizontalWhitespace(WithoutLastNewline.back())) {
+          WithoutLastNewline = WithoutLastNewline.drop_back();
+        } else {
+          break;
+        }
+      }
+
+      if (WithoutLastNewline.ends_with('\\') ||
+          (LangOpts.Trigraphs && WithoutLastNewline.ends_with("??"
+                                                              "/"))) {
```

Comment:
Very weird formatting here (please make "??/" a single string)

Reply:
That gives a "warning: trigraph ignored [-Wtrigraphs]" warning when compiling this

clang-format refuses to put concatenated string literals on the same line.

I've changed this and added a comment, but let me know if this could be done in a cleaner way

Reply:
I mean, can’t you do `"\?\?/"`?

Reply:
Then again, C++17 removed trigraphs, iirc, so just "??/" *should* be fine (you might have to just ignore the fact that clang-format doesn’t like that; we end up having to format something manually every once in a while...)

Reply:
Clang-format doesn't mind `"??/"`, but it's the compiler that complains.

I had no idea `\?` was a valid escape sequence, but apparently it is: https://eel.is/c++draft/lex.ccon#note-1

I'll be using that thanks!

Reply:
> I had no idea `\?` was a valid escape sequence

Yeah, the entire point of it is basically exactly this.